Client/Frontend/Home/ActivityStreamPanel.swift

```diff
@@ -404,6 +404,22 @@ extension ActivityStreamPanel {
 
         let site = self.topSitesManager.content[indexPath.item]
         let eventSource = ASInfo(actionPosition: indexPath.item, source: .topSites)
+        if site.bookmarked == nil {
+            return profile.bookmarks.modelFactory >>== {
+                $0.isBookmarked(site.url)
+                    .uponQueue(dispatch_get_main_queue()) {
+                        guard let isBookmarked = $0.successValue else {
+                            log.error("Error getting bookmark status: \($0.failureValue).")
+                            return
+                        }
+
+                        site.setBookmarked(isBookmarked)
+                        dispatch_async(dispatch_get_main_queue()) {
+                            self.presentContextMenu(site, eventInfo: eventSource, siteImage: siteImage, siteBGColor: siteBGColor)
+                        }
+                }
+            }
+        }
         presentContextMenu(site, eventInfo: eventSource, siteImage: siteImage, siteBGColor: siteBGColor)
```

Comment:
The `return` before the fetch bookmark status call prevents the context menu being called twice, but I'll move it to make that clearer.